### PR TITLE
[Merged by Bors] - feat: Results about convergence of measures of sets assuming convergence of indicator functions.

### PR DESCRIPTION
**Pull request by kkytola**
Adding results about convergence of measures of sets assuming convergence of indicator functions.

---

I could find surprisingly few results about the use of indicators in measure theory, and they were scattered in different places.

The immediate use case for me is the portmanteau theorem (specifically showing that if for all Borel sets with null boundary we have convergence of measures of the set to its measure in the candidate limit, then for all closed sets we have limsup of the measures of the set at most its measure in the candidate limit). For this purpose only the lemma `tendsto_measure_of_tendsto_indicator_of_isFiniteMeasure'` is needed, but the other formulations come with almost no additional effort.

- [x] depends on: #6445 [a.e. measurability for indicators]

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/MeasureTheory/Integral/Indicator.lean
Comment (kkytola) [inline Mathlib/MeasureTheory/Integral/Indicator.lean:9]: If this material is put into a new file, I will write a proper module docstring. But I would like to ask the reviewers to also comment on whether these lemmas should be placed in (various) existing files instead.

File: Mathlib/MeasureTheory/Integral/Indicator.lean
Comment (kkytola) [inline Mathlib/MeasureTheory/Integral/Indicator.lean:92]: Golfing this proof would be very welcome! (Also I am not sure if some variant of the statement would be more helpful instead.)

Comment (ADedecker) [reply Mathlib/MeasureTheory/Integral/Indicator.lean:92]: As I suggested above, I think a first step would be to split out the case where `b` is nonzero. Also, I'd say this is fundamentally a statement about `AEMeasurable`, not `AEStronglyMeasurable`, and it would probably save you some troubles.

Other than that, it's generally best to avoid unfolding the definitions of `AEMeasurable` and `NullMeasurableSet` when you can, and I think [AEMeasurable.nullMeasurable](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/AEMeasurable.html#AEMeasurable.nullMeasurable) allows you to avoid that when getting measurability of the set from measurability of the indicator. For the other direction, I suggest modifying [AEMeasurable.indicator](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/AEMeasurable.html#AEMeasurable.indicator) to work with a null measurable set, using [indicator_eventuallyEq](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Order/Filter/IndicatorFunction.html#indicator_eventuallyEq)

Comment (kkytola) [reply Mathlib/MeasureTheory/Integral/Indicator.lean]: I changed this to about` AEMeasurable`.

Generalizing [AEMeasurable.indicator](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/AEMeasurable.html#AEMeasurable.indicator) to null-measurable instead of measurable sets seems to require generalizing the chain <- [aemeasurable_indicator_iff](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/AEMeasurable.html#aemeasurable_indicator_iff) <- [indicator_ae_eq_restrict](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/MeasureSpace.html#indicator_ae_eq_restrict) <- [piecewise_ae_eq_restrict](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/MeasureSpace.html#piecewise_ae_eq_restrict) <- [ae_restrict_eq](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/MeasureSpace.html#MeasureTheory.ae_restrict_eq) <- [restrict_apply_eq_zero'](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/MeasureSpace.html#MeasureTheory.Measure.restrict_apply_eq_zero') <- [restrict_apply'](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/MeasureSpace.html#MeasureTheory.Measure.restrict_apply') (<- [restrict_toOuterMeasure_eq_toOuterMeasure_restrict](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/MeasureSpace.html#MeasureTheory.Measure.restrict_toOuterMeasure_eq_toOuterMeasure_restrict) <- ...). That might be a worthwhile refactor, but in #6253 I would only need the measurable sets case, so I'd prefer to either keep these more ad hoc proofs or remove the null-measurable and a.e.-measurable versions altogether.

File: Mathlib/MeasureTheory/Integral/Indicator.lean
Comment (kkytola) [reply Mathlib/MeasureTheory/Integral/Indicator.lean]: I agree, this is nicer. I made everything assume `[MeasurableSingletonClass]` and `[NeZero]`.

File: Mathlib/MeasureTheory/Integral/Indicator.lean
Comment (ADedecker) [inline Mathlib/MeasureTheory/Integral/Indicator.lean:126]:
```suggestion
  simp_rw [← measurable_indicator_const_iff _ (1 : ℝ≥0∞)] at As_mble ⊢
  exact measurable_of_tendsto_ennreal' L As_mble h_lim
```

File: Mathlib/MeasureTheory/Integral/Indicator.lean
Comment (ADedecker) [inline Mathlib/MeasureTheory/Integral/Indicator.lean:141]:
```suggestion
  simp_rw [← aeMeasurable_indicator_const_iff _ μ (1 : ℝ≥0∞)] at As_mble ⊢
  exact aemeasurable_of_tendsto_metrizable_ae L As_mble h_lim
```

File: Mathlib/MeasureTheory/Integral/Indicator.lean
Comment (ADedecker) [inline Mathlib/MeasureTheory/Integral/Indicator.lean:146]:
```suggestion
lemma tendsto_measure_of_ae_tendsto_indicator (μ : Measure α) (A_mble : MeasurableSet A)
```
Same below

File: Mathlib/MeasureTheory/Integral/Indicator.lean
Comment (kkytola) [reply Mathlib/MeasureTheory/Integral/Indicator.lean]: Thank you very much, this is great! I also realized that contrary to what I said above, I probably do want the a.e.-measurable/null-measurable versions after all, so I will be extremely happy if they are not removed from the PR.

I don't know if I marked the "depends on #6445" correctly, since this PR didn't become automatically marked "blocked by another PR". Do I need to add that label by hand when I add a "depends on" not at the PR creation but later?

Comment (ADedecker) [reply Mathlib/MeasureTheory/Integral/Indicator.lean]: If [this message](https://github.com/leanprover-community/mathlib4/pull/6225#issuecomment-1670885097) appears then it's correctly ported. I think the bot is supposed to add the label by itself, but sometimes it takes some time / doesn't work.

File: Mathlib/MeasureTheory/Integral/Indicator.lean
Comment (kkytola) [reply Mathlib/MeasureTheory/Integral/Indicator.lean:9]: I think I will be able to move these to rather natural choices of existing files. The results of the same flavor will then be somewhat scattered, but they do rely on different definitions (measurability vs a.e.-meaurability, null-measurability, etc.), so that seems ok to me, at least.

Comment (ADedecker) [reply Mathlib/MeasureTheory/Integral/Indicator.lean:9]: Yes that sounds good

Comment (kkytola) [reply Mathlib/MeasureTheory/Integral/Indicator.lean:9]: Actually, for the two remaining results in the new file `Indicator.lean`, I could not find natural places. The import-prerequisites are `Mathlib.MeasureTheory.Constructions.BorelSpace.Metrizable` and
`Mathlib.MeasureTheory.Integral.Lebesgue`.

Were there tools for this besides `#find_home`? That tool unfortunately reports `Mathlib.MeasureTheory.Integral.Indicator`, i.e., the new file itself.

Besides these two lemmas, the others are now scattered in different files in a way that I believe is quite natural.

Comment (kkytola) [reply Mathlib/MeasureTheory/Integral/Indicator.lean:9]: Of course one could argue that these lemmas do almost nothing but combine the `tendsto_measure_of_ae_tendsto_indicator` with preservation of measurability in limits. But I think for good API, it is  nicer not to request the user to explicitly provide that trivial preservation of measurability. That is why I think we should have the remaining two lemmas.

**Comment by RemyDegenne** [comment]
Are all the conversations above resolved? If they are, could you mark them accordingly?
@ADedecker, does the PR look good to you now?

**Comment by ADedecker** [comment]
Oh sorry this went out of my head. I just had a quick look and everything looks good to me (and I believe all the conversations are indeed resolved). The only thing I'm not sure about is whether the measure arguments should be implicit/use `volume_tac`, but you probably know the convention better than me so I'll let you comment on it.

File: Mathlib/MeasureTheory/Integral/Lebesgue.lean
Comment (RemyDegenne) [inline Mathlib/MeasureTheory/Integral/Lebesgue.lean:2119]:
```suggestion
lemma tendsto_measure_of_ae_tendsto_indicator {μ : Measure α} (A_mble : MeasurableSet A)
```

File: Mathlib/MeasureTheory/Integral/Lebesgue.lean
Comment (RemyDegenne) [inline Mathlib/MeasureTheory/Integral/Lebesgue.lean:2149]:
```suggestion
lemma tendsto_measure_of_tendsto_indicator [NeBot L] {μ : Measure α}
```

**Review by RemyDegenne — APPROVED** [review]
It looks like some of the measure arguments should be implicit, but otherwise it all looks good to me.
bors d+

**Comment by kkytola** [comment]
> It looks like some of the measure arguments should be implicit, but otherwise it all looks good to me. bors d+

Thanks!

I will make these implicit. Probably the reason I write more explicit measure arguments than the other parts of the library is that I use these for weak convergence of measures where there are always multiple measures and so it is often necessary to specify the measure. But with Lean4's named arguments, this should be smooth with implicit measures as well.

**Comment by kkytola** [comment]
> ✌️ kkytola can now approve this pull request. To approve and merge a pull request, simply reply with `bors r+`. More detailed instructions are available [here](https://bors.tech/documentation/getting-started/#reviewing-pull-requests).

bors r+

**Comment by fpvandoorn** [comment]
@kkytola This PR has added a **huge** import to `Integral/Lebesgue`: a file that imports a significant part of the analysis library.

Could you please undo that part of the PR (maybe move the new lemmas to a new file) and add a line `assert_not_exists NormedSpace` to make sure we don't add such a big import to this file again? Thanks!

**Comment by kkytola** [comment]
> @kkytola This PR has added a _**huge**_ import to `Integral/Lebesgue`: a file that imports a significant part of the analysis library.
> 
> Could you please undo that part of the PR (maybe move the new lemmas to a new file) and add a line `assert_not_exists NormedSpace` to make sure we don't add such a big import to this file again? Thanks!

PR #6850 creates a new file for two lemmas and undoes the adding of the import.

In the original PR and review, the creation of a new file vs. adding an import [was discussed from the start](https://github.com/leanprover-community/mathlib4/pull/6225#discussion_r1279767582). To me, these are still complicated trade-offs. I definitely understand the goal to minimize imports for efficiency and even mathematical elegance, but how to weigh that against creation of new files and scattering thematically related lemmas to different places is not easy for me. Sorry about having created an unsatisfactory situation, I hope the fix PR makes it better!

**Comment by ADedecker** [comment]
Ah yes sorry this is my fault, I suggested this move. I didn't realize importing `borel` was a heavy import for measure theory (maybe Bourbaki bias here...). @fpvandoorn do you think it would make sense to split off the `NormedAddCommGroup` lemmas from [Mathlib.MeasureTheory.Constructions.BorelSpace.Basic](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Constructions/BorelSpace/Basic.html)? Would that be enough to make this import viable again?

**Comment by fpvandoorn** [comment]
Thanks for the new PR, and no worries, these things can happen!

Yes, I was also wondering if the problem was somewhere else.

If easy it might be nice if [`MeasureTheory.Constructions.BorelSpace.Basic`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Constructions/BorelSpace/Basic.html#top) doesn't import [`Mathlib.Analysis.Normed.Group.Basic`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Normed/Group/Basic.html). But in this case the relevant thing is that [`MeasureTheory.Constructions.BorelSpace.Metrizable`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Constructions/BorelSpace/Metrizable.html#top) has the much heavier import [`Mathlib.Topology.MetricSpace.Metrizable`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/MetricSpace/Metrizable.html#top) since that imports [`Mathlib.Analysis.SpecificLimits.Basic`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/SpecificLimits/Basic.html). Somewhere this chain should probably be broken.

I also added a comment to #6850 that probably allows us to make the lemmas added in this PR more elementary.

**Pull request closed without merging**
